docsrc/config.rst
Configuration file
==================

``luacheck`` tries to load configuration from ``.luacheckrc`` file in the current directory. If not found, it will look for it in the parent directory and so on, going up until it reaches file system root. Path to config can be set using ``--config`` option, in which case it will be used during recursive loading. Paths within config are interpreted relatively to the directory from which it was loaded.

Config loading can be disabled using ``--no-config`` flag.

If neither of ``--config``, ``--no-config``, and ``--no-default-config`` options are used, ``luacheck`` will attempt to load configuration from value of ``--default-config`` option,
or ``%LOCALAPPDATA%\Luacheck\.luacheckrc`` on Windows, ``~/Library/Application Support/Luacheck/.luacheckrc`` on OS X/macOS, and ``$XDG_CONFIG_HOME/luacheck/.luacheckrc``
or ``~/.config/luacheck/.luacheckrc`` on other systems by default. Paths within default config are interpreted relatively to the current directory.

Config is simply a Lua script executed by ``luacheck``. It may set various options by assigning to globals or by returning a table with option names as keys.

Options loaded from config have the lowest priority: it's possible to overwrite them with CLI options or inline options.

.. _options:

Config options
--------------

============================= ======================================== ===================
Option                        Type                                     Default value
============================= ======================================== ===================
``quiet``                     Integer in range 0..3                    ``0``
``color``                     Boolean                                  ``true``
``codes``                     Boolean                                  ``false``
``ranges``                    Boolean                                  ``false``
``formatter``                 String or function                       ``"default"``
``cache``                     Boolean or string                        ``false``
``jobs``                      Positive integer                         ``1``
``exclude_files``             Array of strings                         ``{}``
``include_files``             Array of strings                         (Include all files)
``global``                    Boolean                                  ``true``
``unused``                    Boolean                                  ``true``
``redefined``                 Boolean                                  ``true``
``unused_args``               Boolean                                  ``true``
``unused_secondaries``        Boolean                                  ``true``
``self``                      Boolean                                  ``true``
``std``                       String or set of standard globals        ``"max"``
``globals``                   Array of strings or field definition map ``{}``
``new_globals``               Array of strings or field definition map (Do not overwrite)
``read_globals``              Array of strings or field definition map ``{}``
``new_read_globals``          Array of strings or field definition map (Do not overwrite)
``not_globals``               Array of strings                         ``{}``
``operators``                 Array of strings                         ``{}``
``compat``                    Boolean                                  ``false``
``allow_defined``             Boolean                                  ``false``
``allow_defined_top``         Boolean                                  ``false``
``module``                    Boolean                                  ``false``
``max_line_length``           Number or ``false``                      ``120``
``max_code_line_length``      Number or ``false``                      ``120``
``max_string_line_length``    Number or ``false``                      ``120``
``max_comment_line_length``   Number or ``false``                      ``120``
``max_cyclomatic_complexity`` Number or ``false``                      ``false``
``ignore``                    Array of patterns (see :ref:`patterns`)  ``{}``
``enable``                    Array of patterns                        ``{}``
``only``                      Array of patterns                        (Do not filter)
============================= ======================================== ===================

An example of a config which makes ``luacheck`` ensure that only globals from the portable intersection of Lua 5.1, Lua 5.2, Lua 5.3, Lua 5.4, Lua 5.5 and LuaJIT 2.0 are used, as well as disables detection of unused arguments:

.. code-block:: lua
   :linenos:

   std = "min"
   ignore = {"212"}

.. _custom_stds:


Custom sets of globals
----------------------

``std`` option allows setting a custom standard set of globals using a table. This table can have two fields: ``globals`` and ``read_globals``.
Both of them should contain a field definition map defining some globals. The simplest way to define globals is to list their names:

.. code-block:: lua
   :linenos:

   std = {
      globals = {"foo", "bar"}, -- these globals can be set and accessed.
      read_globals = {"baz", "quux"} -- these globals can only be accessed.
   }

For globals defined like this Luacheck will additionally consider any fields within them defined. To define a global with a restricted set of fields, use
global name as key and a table as value. In that table, ``fields`` subtable can contain the fields in the same format:

.. code-block:: lua
   :linenos:

   std = {
      read_globals = {
         foo = { -- Defining read-only global `foo`...
            fields = {
               field1 = { -- `foo.field1` is now defined...
                  fields = {
                     nested_field = {} -- `foo.field1.nested_field` is now defined...
                  }
               },
               field2 = {} -- `foo.field2` is defined.
            }
         }
      }
   }

Globals and fields can be marked read-only or not using ``read_only`` property with a boolean value.
Property ``other_fields`` controls whether the global or field can also contain other unspecified fields:

.. code-block:: lua
   :linenos:

   std = {
      read_globals = {
         foo = { -- `foo` and its fields are read-only by default (because they are within `read_globals` table).
            fields = {
               bar = {
                  read_only = false, -- `foo.bar` is not read-only, can be set.
                  other_fields = true, -- `foo.bar[anything]` is defined and can be set or mutated (inherited from `foo.bar`).
                  fields = {
                     baz = {read_only = true}, -- `foo.bar.baz` is read-only as an exception.
                  }
               }
            }
         }
      }
   }

Custom sets can be given names by mutating global ``stds`` variable, so that they can then be used in ``--std`` CLI option
and ``std`` inline and config option.

.. code-block:: lua
   :linenos:

   stds.some_lib = {...}
   std = "min+some_lib"

In config, ``globals``, ``new_globals``, ``read_globals``, and ``new_read_globals`` can also contain definitions in same format:

.. code-block:: lua
   :linenos:

   read_globals = {
      server = {
         fields = {
            -- Allow mutating `server.sessions` with any keys...
            sessions = {read_only = false, other_fields = true},
            -- other fields...
         }
      },
      --- other globals...
   }

Per-file and per-path overrides
-------------------------------

The environment in which ``luacheck`` loads the config contains a special global ``files``. When checking a file ``<path>``, ``luacheck`` will override options from the main config with entries from ``files[<glob>]`` if ``<glob>`` matches ``<path>``, applying entries for more general globs first. For example, the following config re-enables detection of unused arguments only for files in ``src/dir``, but not for files ending with ``_special.lua``:

.. code-block:: lua
   :linenos:

   std = "min"
   ignore = {"212"}
   files["src/dir"] = {enable = {"212"}}
   files["src/dir/**/*_special.lua"] = {ignore = {"212"}}

Note that ``files`` table supports autovivification, so that

.. code-block:: lua

   files["src/dir"].enable = {"212"}

and

.. code-block:: lua

   files["src/dir"] = {enable = {"212"}}

are equivalent.

The configs are processed in order of increasing specificity. ``globals`` and ``read_globals`` will add to the set of allowed globals. ``not_globals`` can be used to remove previously allowed globals; ``new_globals`` and ``new_read_globals`` can be used to override the set of globals, wiping all previously allowed ``globals`` or ``read_globals`` respectively, and replacing them with new entries.

Default per-path std overrides
------------------------------

``luacheck`` uses a set of default per-path overrides:

.. code-block:: lua
   :linenos:

   files["**/spec/**/*_spec.lua"].std = "+busted"
   files["**/test/**/*_spec.lua"].std = "+busted"
   files["**/tests/**/*_spec.lua"].std = "+busted"
   files["**/*.rockspec"].std = "+rockspec"
   files["**/*.luacheckrc"].std = "+luacheckrc"
   files["**/config.ld"].std = "+ldoc"

These are added to the global ``std`` specified in the config file.
Each of these can be overriden by setting a different ``std`` value for the corresponding key in ``files``.
Setting ``std`` on the commandline removes these default overrides.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.2.0] - 2024-05-24

### Features

- Add builtin standards support for minetest (#108) — @BuckarooBanzay and @appgurueu

### Performance

- Memoize results to addresses speed refression from new feature in v0.26 (#105) — @tomlau10

## [1.1.2] - 2023-12-08

### Features

- Support NO_COLOR environment variable (#102) — @ligurio

### Bug Fixes

- Update SILE builtin with more allowed variables — @alerque

## [1.1.1] - 2023-06-09

### Features

- Update Löve standard to 11.4 (#95) — @RunningDroid
- Documentation improvements (#92 and #89) — @rcloran and @hramrach

### Bug Fixes

- Correct compound operators to not crash on modifying upvalues (#96) — @arichard4
- Fix warning 582 (error prone negation) not applying to subexpressions (#94) — @appgurueu

## [1.1.0] - 2022-12-19

### Features

- Add builtin rule set for SILE globals (#79) — @alerque
- Implement support for compound operators (#66) — @a2 & @arichard4

### Bug Fixes

- Correct circular reference detection visavis OpSet — @arichard4
- Remove unnecessary symbol from Playdate std (#84) — @DidierMalenfant

## v1.0.0 (2022-08-24)

### Documentation

- [**breaking**] Follow semver guidelines, next release will be v1.x.y

### Features

- Overhaul docker container to run on Lua 5.4 — @alerque
- Store cached luacheck values per-version in case of changes — @arichard4
- Set_default_std for ldoc — @Aire-One
- Add builtin std option for Ldoc globals — @Aire-One
- Add builtin std option for the Playdate SDK — @DidierMalenfant

## v0.26.1 (2022-04-23)

### Bug Fixes

- Exempt special builtin \_ENV from 214 warning

### Features

- In case of no home environment, default to caching in CWD (#60)
- Add multi-thread support to container (#59)

### Miscellaneous Tasks

- Tweak warning message for 214 to be more explicit

## v0.26.0 (2022-03-23)

### New Features

* Add "unused hint" flag — @javierguerragiraldez
  Function arguments that start with a single underscore get an "unused hint".
  Leaving them unused doesn't result in a warning.
  Using them, on the other hand, is a new warning (№ 214).
* Add hook to enable Luacheck for use in pre-commit — @mblayman
* Warn on error-prone and unnecessary negations — @arichard4
  Two new warnings (№ 581 and 582) flag error-prone operator orders.
* Add Dockerfile implementation — @MartinBroers
  A 6.34MB containerized image with everything needed to run the linter.
  Build your own or pull prebuilt images from GHRC.
* Setup repository for use *as* a GitHub Action — @alerque
  Lint your repositories using GitHub Action workflows with just a single `uses:` step.

### Fixes

* Don't mark variables as accessed if only references are circular — @arichard4
* Make test suite Lua 5.4 compatible — @alerque
* Correct small issues in documentation — various

### Miscellaneous

* Overhaul CI workflows for testing, linting, building, and releasing — @alerque
* Update URLs and documentation reflecting new repository home — @lunarmodules

## 0.25.0 (2020-08-25)

### New features

* New values for CLI option --std: lua54, lua54c
* New lua54 standard library definitions

### Fixes

* Lua 5.4 allows 31bits utf8 codepoint

## 0.24.0 (2020-08-20)

### Changes

* Caching now uses files in a global directory instead of local
  `.luacheckcache` file. Default cache directory is
  `%LOCALAPPDATA%\Luacheck\Cache` on Windows,
  `~/Library/Caches/Luacheck` on OS X/macOS, and
  `$XDG_CACHE_HOME/luacheck` or `~/.config/luacheck` on other systems.

### Fixes

* Fixes for using Luacheck with Lua 5.4:
  - Parse 5.4 attributes in `local` declarations (but ignores them for now)
  - Luacheck itself can also run with Lua 5.4
* Fixed `randomize` missing from `busted` set of standard globals (#183).
* Added additional `table` and `thread` definitions for `ngx_lua`.
* Added additional `match` definition for `busted`.

### Miscellaneous

* Upgraded Windows binary components: Lua 5.3.4 -> 5.3.5,
  LuaFileSystem 1.6.3 -> 1.7.0.

## 0.23.0 (2018-09-18)

### Breaking changes

* Removed `--no-inline` CLI option and `inline` config option, inline options
  are now always enabled.
* Inline comments are now supposed to be only in short comments
  but not long ones.
* Installer script (install.lua) is removed. Luacheck can still be installed
  manually by recursively copying `src/*` to a directory in `package.path`
  and copying `bin/luacheck.lua` to a directory in `PATH` as `luacheck`.

### New features and improvements

* Warning columns are now reported in Unicode codepoints if input is
  valid UTF-8 (#45).
* Added indentaion-based guessing of a better location for missing `end`
  and `until` syntax errors.
* Added `luacheckrc` set of allowed globals containing globals used in
  Luacheck config to set options.
* Added default stds equivalent to predefined per-path std overrides
  in config:
  - `files["**/spec/**/*_spec.lua"].std = "+busted"`;
  - `files["**/test/**/*_spec.lua"].std = "+busted"`;
  - `files["**/tests/**/*_spec.lua"].std = "+busted"`;
  - `files["**/*.rockspec"].std = "+rockspec"`;
  - `files["**/*.luacheckrc"].std = "+luacheckrc"`.
* Added detection of numeric for loops going from `#t` to `1` without
  negative step (#160).
* Added support for LuaRocks 3 module autodetection when checking
  rockspecs (#176).
* Updated `love` standard for LÖVE 11.1 (#178).

### Changes

* Default set of standard globals is now always `max`, allowing globals of all
  Lua versions. `_G` std is deprecated.

### Fixes

* Added missing globals to `rockspec` std: `hooks`, `deploy`,
  `build_dependencies`, `test_dependencies`, and `test`.
* Fixed line lengths appearing in the output before other warnings on the same
  line even if their column numbers are smaller.

### Miscellaneous

* Luacheck now depends on argparse instead of bundling it.
* LuaFileSystem dependency is now required.

## 0.22.1 (2018-07-01)

### Improvements


src/luacheck/options.lua
local options = {}

local builtin_standards = require "luacheck.builtin_standards"
local standards = require "luacheck.standards"
local utils = require "luacheck.utils"

local boolean = utils.has_type("boolean")
local number_or_false = utils.has_type_or_false("number")
local array_of_strings = utils.array_of("string")

-- Validates std string.
-- Returns an array of std names with `add` field if there is `+` at the beginning of the string.
-- On validation error returns `nil` and an error message.
local function split_std(std, stds)
   local parts = utils.split(std, "+")

   if parts[1]:match("^%s*$") then
      parts.add = true
      table.remove(parts, 1)
   end

   for i, part in ipairs(parts) do
      parts[i] = utils.strip(part)

      if not stds[parts[i]] then
         return nil, ("unknown std '%s'"):format(parts[i])
      end
   end

   return parts
end

local function std_or_array_of_strings(x, stds)
   if type(x) == "string" then
      local ok, err = split_std(x, stds)
      return not not ok, err
   elseif type(x) == "table" then
      return standards.validate_std_table(x)
   else
      return false, "string or table expected, got " .. type(x)
   end
end

local function field_map(x)
   if type(x) == "table" then
      return standards.validate_globals_table(x)
   else
      return false, "table expected, got " .. type(x)
   end
end

options.nullary_inline_options = {
   global = boolean,
   unused = boolean,
   redefined = boolean,
   unused_args = boolean,
   unused_secondaries = boolean,
   self = boolean,
   compat = boolean,
   allow_defined = boolean,
   allow_defined_top = boolean,
   module = boolean
}

options.variadic_inline_options = {
   globals = field_map,
   read_globals = field_map,
   new_globals = field_map,
   new_read_globals = field_map,
   not_globals = array_of_strings,
   ignore = array_of_strings,
   enable = array_of_strings,
   only = array_of_strings,
}

options.all_options = {
   std = std_or_array_of_strings,
   max_line_length = number_or_false,
   max_code_line_length = number_or_false,
   max_string_line_length = number_or_false,
   max_comment_line_length = number_or_false,
   max_cyclomatic_complexity = number_or_false,
   operators = array_of_strings
}

utils.update(options.all_options, options.nullary_inline_options)
utils.update(options.all_options, options.variadic_inline_options)

-- Returns true if opts is valid option_set or is nil.
-- Otherwise returns false and an error message.
function options.validate(option_set, opts, stds)
   if opts == nil then
      return true
   end

   if type(opts) ~= "table" then
      return false, "option table expected, got " .. type(opts)
   end

   stds = stds or builtin_standards

   for option, validator in utils.sorted_pairs(option_set) do
      if opts[option] ~= nil then
         local ok, err = validator(opts[option], stds)

         if not ok then
            return false, ("invalid value of option '%s': %s"):format(option, err)
         end
      end
   end

   return true
end

-- Option stack is an array of options with options closer to end
-- overriding options closer to beginning.

-- Extracts sequence of active std tables from an option stack.
local function get_std_tables(opts_stack, stds)
   local base_std
   local add_stds = {}
   local no_compat = false

   for _, opts in utils.ripairs(opts_stack) do
      if opts.compat and not no_compat then
         base_std = stds.max
         break
      elseif opts.compat == false then
         no_compat = true
      end

      if opts.std then
         if type(opts.std) == "table" then
            base_std = opts.std
            break
         else
            local parts = split_std(opts.std, stds)

            for _, part in ipairs(parts) do
               table.insert(add_stds, stds[part])
            end

            if not parts.add then
               base_std = {}
               break
            end
         end
      end
   end

   table.insert(add_stds, 1, base_std or stds.max)
   return add_stds
end

-- Returns index of the last option table in a stack that uses given option,
-- or zero if the option isn't used anywhere.
local function index_of_last_option_usage(opts_stack, option_name)
   for index, opts in utils.ripairs(opts_stack) do
      if opts[option_name] then
         return index
      end
   end

   return 0
end

local function split_field(field_name)
   return utils.split(field_name, "%.")
end

local function field_comparator(field1, field2)
   local parts1 = field1[1]
   local parts2 = field2[1]

   for i = 1, math.max(#parts1, #parts2) do
      local part1 = parts1[i]
      local part2 = parts2[i]

      if not part1 then
         return true
      elseif not part2 then
         return false
      end

      if part1 ~= part2 then
         return part1 < part2
      end
   end

   return false
end

-- Combine all stds and global related options into one final definition table.
-- A definition table may have fields `read_only` (boolean), `other_fields` (boolean),
-- and `fields` (maps field names to definition tables).
-- Std table format is similar, except at the top level there are two fields
-- `globals` and `read_globals` mapping to top-level field tables. Also in field tables
-- it's possible to use field names in array part as a shortcut:
-- `{fields = {"foo"}}` is equivalent to `{fields = {foo = {}}}` or `{fields = {foo = {other_fields = true}}}`
-- in top level fields tables.
local function get_final_std(opts_stack, stds)
   local final_std = {}
   local std_tables = get_std_tables(opts_stack, stds)

   for _, std_table in ipairs(std_tables) do
      standards.add_std_table(final_std, std_table)
   end

   local last_new_globals = index_of_last_option_usage(opts_stack, "new_globals")
   local last_new_read_globals = index_of_last_option_usage(opts_stack, "new_read_globals")

   for index, opts in ipairs(opts_stack) do
      local globals = (index >= last_new_globals) and (opts.new_globals or opts.globals)
      local read_globals = (index >= last_new_read_globals) and (opts.new_read_globals or opts.read_globals)

      local new_fields = {}

      if globals then
         for _, g in ipairs(globals) do
            table.insert(new_fields, {split_field(g), false})
         end
      end

      if read_globals then
         for _, read_global in ipairs(read_globals) do
            table.insert(new_fields, {split_field(read_global), true})
         end
      end

      if globals and read_globals then
         -- If there are both globals and read-only globals defined in one options table,
         -- it's important that more general definitions are applied first,
         -- otherwise they will completely overwrite more specific definitions.
         -- E.g. `globals x` should be applied before `read globals x.y`.
         table.sort(new_fields, field_comparator)
      end

      for _, field in ipairs(new_fields) do
         standards.overwrite_field(final_std, field[1], field[2])
      end

      standards.add_std_table(final_std, {globals = globals, read_globals = read_globals}, true, true)

      if opts.not_globals then
         for _, not_global in ipairs(opts.not_globals) do
            standards.remove_field(final_std, split_field(not_global))
         end
      end
   end

   standards.finalize(final_std)
   return final_std
end

local function get_scalar_opt(opts_stack, option, default)
   for _, opts in utils.ripairs(opts_stack) do
      if opts[option] ~= nil then
         return opts[option]
      end
   end

   return default
end

local line_length_suboptions = {"max_code_line_length", "max_string_line_length", "max_comment_line_length"}

local function get_max_line_opts(opts_stack)
   local res = {max_line_length = 120}

   for _, opt_name in ipairs(line_length_suboptions) do
      res[opt_name] = res.max_line_length
   end

   for _, opts in ipairs(opts_stack) do
      if opts.max_line_length ~= nil then
         res.max_line_length = opts.max_line_length

         for _, opt_name in ipairs(line_length_suboptions) do
            res[opt_name] = opts.max_line_length
         end
      end

      for _, opt_name in ipairs(line_length_suboptions) do
         if opts[opt_name] ~= nil then
            res[opt_name] = opts[opt_name]
         end
      end
   end

   return res
end

local function anchor_pattern(pattern, only_start)
   if not pattern then
      return
   end

   if pattern:sub(1, 1) == "^" or pattern:sub(-1) == "$" then
      return pattern
   else
      return "^" .. pattern .. (only_start and "" or "$")
   end
end

-- Returns {pair of normalized patterns for code and name}.
-- `pattern` can be:
--    string containing '/': first part matches warning code, second - variable name;
--    string containing letters: matches variable name;
--    otherwise: matches warning code.
-- Unless anchored by user, pattern for name is anchored from both sides
-- and pattern for code is only anchored at the beginning.
local function normalize_pattern(pattern)
   local code_pattern, name_pattern
   local slash_pos = pattern:find("/")

   if slash_pos then
      code_pattern = pattern:sub(1, slash_pos - 1)
      name_pattern = pattern:sub(slash_pos + 1)
   elseif pattern:find("[_a-zA-Z]") then
      name_pattern = pattern
   else
      code_pattern = pattern
   end

   return {anchor_pattern(code_pattern, true), anchor_pattern(name_pattern)}
end

-- From most specific to less specific, pairs {option, pattern}.
-- Applying macros in order is required to get deterministic results
-- and get sensible results when intersecting macros are used.
-- E.g. unused = false, unused_args = true should leave unused args enabled.
local macros = {
   {"unused_args", "21[23]"},
   {"global", "1"},
   {"unused", "[23]"},
   {"redefined", "4"}
}

-- Returns array of rules which should be applied in order.
-- A rule is a table {{pattern*}, type}.
-- `pattern` is a non-normalized pattern.
-- `type` can be "enable", "disable" or "only".
local function get_rules(opts_stack)
   local rules = {}
   local used_macros = {}

   for _, opts in utils.ripairs(opts_stack) do
      for _, macro_info in ipairs(macros) do
         local option, pattern = macro_info[1], macro_info[2]

         if not used_macros[option] then
            if opts[option] ~= nil then
               table.insert(rules, {{pattern}, opts[option] and "enable" or "disable"})
               used_macros[option] = true
            end
         end
      end

      if opts.ignore then
         table.insert(rules, {opts.ignore, "disable"})
      end

      if opts.only then
         table.insert(rules, {opts.only, "only"})
      end

      if opts.enable then
         table.insert(rules, {opts.enable, "enable"})
      end
   end

   return rules
end

local function normalize_patterns(rules)
   local res = {}

   for i, rule in ipairs(rules) do
      res[i] = {{}, rule[2]}

      for j, pattern in ipairs(rule[1]) do
         res[i][1][j] = normalize_pattern(pattern)
      end
   end

   return res
end

local function get_operators(opts_stack)
   local operators, operatorsMap = nil, nil

   for _, opts in ipairs(opts_stack) do
      if opts.operators then
         operators = operators or {}
         operatorsMap = operatorsMap or {}

         for _, op in ipairs(opts.operators) do
            if not operatorsMap[op] then
               table.insert(operators, op)
               operatorsMap[op] = true
            end
         end
      end
   end

   return operators
end

local scalar_options = {
   unused_secondaries = true,
   self = true,
   module = false,
   allow_defined = false,
   allow_defined_top = false,
   max_cyclomatic_complexity = false
}

-- Returns normalized options.
-- Normalized options have fields:
--    std: normalized std table, see `luacheck.standards` module;
--    unused_secondaries, self, module, allow_defined, allow_defined_top: booleans;
--    max_line_length: number or false;
--    rules: see get_rules.
function options.normalize(opts_stack, stds)
   local res = {}
   stds = stds or builtin_standards
   res.std = get_final_std(opts_stack, stds)
   res.operators = get_operators(opts_stack)

   for option, default in pairs(scalar_options) do
      res[option] = get_scalar_opt(opts_stack, option, default)
   end

   local max_line_opts = get_max_line_opts(opts_stack)
   utils.update(res, max_line_opts)
   res.rules = normalize_patterns(get_rules(opts_stack))

   return res
end

return options
